### DISCUSSION TITLE
Wy does the libvips mixin "black" to the colors that are fading into transparent background?

### DISCUSSION THREAD

**Discussion by hanssonrickard** [libvips/libvips#2173] [Q&A] [answered]
Hi.
A bit strange to explain, but will post a screenshot and if you need i can post the actual test images as well.
I have a test tif that i created in Photoshop that has a transparent background and just some hand drawn lines in it, in two different colors. The edges of the drawn lines are "fuzzy" going from opaque white in the middle of the line to fuzzy edges that are fading into transparent.

If i open the file in libvips via pyvips and save it again as tif, png and jpg. The transparent background is intact, but the fuzzy edges of the lines are now mixed with what appears to be black/grey color, not keeping their "paint color -> transp" fade any more. 
For the jpg output i added a "flatten" command via pyvips" and set the background to white for the flatten command (tested it also for tif and png output, same results) and the background is turned into white (instad of standard black), but i still get the different colored fuzzy edges of the lines.

The simple pyvips code i have tested is (flatten tested on and off):

	image = pyvips.Image.new_from_file(input_file, access='sequential')
	#image = image.flatten(background=[255, 255, 255])
	image = image.icc_transform(output_icc_profile, embedded=True, input_profile=default_input_profile)
	image.write_to_file(tmpdir + '/' + input_name_no_ext + output_extension)
	
Am i missing something here to get "correct" result?
Here i screenshot as well where all four files are seen, original bottom right as well as the three resulting images.

<img width="1874" alt="screenshot" src="https://user-images.githubusercontent.com/4253602/112897770-4f279400-90e0-11eb-982b-7322c591e448.png">

**Comment by hanssonrickard** [comment]
Hmm, could it be that the libvips version that is part of Debian 9 packages is quote old. 
I could see it is vips 8.4.5 and it feels quite old compare to what is available.

**Comment by jcupitt** [comment]
Yes, you could need a newer libvips -- yours is five years old. If you post your test images I could try on master for you.

**Reply by hanssonrickard** [reply]
Here is the original test tif.
What we are doing is looking into replacing imagemagick with something faster and files with transparency is something used by our print customers a lot.

Can libvips also create for example a checker-pattern as background, for example if we wish to create jpg previews that does not support transparency and we wish to sindicate/simulate it with a pattern, is that possible?

[transparent_orig.tif.zip](https://github.com/libvips/libvips/files/6228223/transparent_orig.tif.zip)

**Reply by hanssonrickard** [reply]
Oh, by the way, the libvips i am using is from the Debian repository and it is a shame that it is so old in their repo.
But that is not your concern, just a shame it is not better up2date.

**Comment by jcupitt** [comment, marked as answer]
I tried:

```
vips icc_transform transparent_orig.tif x.png srgb
```

To make:

![x](https://user-images.githubusercontent.com/580843/112975598-c21a2480-914b-11eb-94e8-5297873c73b5.png)

Obviously the white star doesn't show up too well, but there are no stray shadows.

**Comment by jcupitt** [comment]
I made a background tile:

![tile](https://user-images.githubusercontent.com/580843/112976633-0659f480-914d-11eb-918e-4f9db1334053.png)

You can composite things onto that like this:

```python
#!/usr/bin/python3

import sys
import pyvips

tile = pyvips.Image.new_from_file("tile.png")

rgba = pyvips.Image.new_from_file(sys.argv[1], access="sequential")
bg = tile \
    .replicate(1 + rgba.width / tile.width, 1 + rgba.height / tile.height) \
    .crop(0, 0, rgba.width, rgba.height)
rgb = bg.composite(rgba, "over")[0:3]

rgb.write_to_file(sys.argv[2])
```

If I run like this:

```
./tile-background.py x.png x.jpg
```

I get:

![x](https://user-images.githubusercontent.com/580843/112976776-30abb200-914d-11eb-8eda-9c252a4fef2e.jpg)

**Reply by hanssonrickard** [reply]
Ok, thanks for both replies.
So, most probably a problem related to the older version i am using.
I will see if i can setup a VM with Debian 9 and see if i can create a new libvips version compatible with debian 9.
Thanks for the example for a check pattern. Much appreciated.

Thanks for the test and time you took to reply to this thread.

**Reply by jcupitt** [reply]
I've used docker for this kind of thing, fwiw. Sample dockerfile for php-vips on debian9:

https://github.com/jcupitt/docker-builds/blob/master/php-vips-debian9/Dockerfile

You'd want to swap php for py, of course.

**Reply by hanssonrickard** [reply]
Ah, that is a great idea.
Thanks.
Will give it a go.
